alf/algorithms/diayn_algorithm.py

```diff
@@ -0,0 +1,135 @@
+# Copyright (c) 2020 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import namedtuple
+
+import gin.tf
+import tensorflow as tf
+
+from tf_agents.networks.network import Network
+import tf_agents.specs.tensor_spec as tensor_spec
+
+from alf.algorithms.algorithm import Algorithm, AlgorithmStep, LossInfo
+from alf.utils.encoding_network import EncodingNetwork
+
+DIAYNInfo = namedtuple("DIAYNInfo", ["reward", "loss"])
+
+
+@gin.configurable
+class DIAYNAlgorithm(Algorithm):
+    """Diversity is All You Need Module
+
+    This module learns a set of skill-conditional policies in an unsupervised way.
+
+    See Eysenbach et al "Diversity is All You Need: Learning Diverse Skills without a Reward Function"
+    """
+
+    def __init__(self,
+                 num_of_skills,
+                 feature_spec,
+                 hidden_size=256,
+                 encoding_net: Network = None,
+                 discriminator_net: Network = None,
+                 name="DIAYNAlgorithm"):
+        """Create a DIAYNAlgorithm.
+
+        Args:
+            num_of_skills: number of skills
+            hidden_size (int|tuple): size of hidden layer(s)
+            encoding_net (Network): network for encoding observation into a
+                latent feature specified by feature_spec. Its input is same as
+                the input of this algorithm.
+            discriminator_net (Network): network for predicting the skill labels based on the observation.
```

Comment:
80c per line. You can set editor.rulers in vscode settings:
{
"editor.rulers": [80]
}